### PR TITLE
H-4827: Allow direct conversion from policies to filters

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

This PR refactors the policy evaluation system to enable direct conversion from Cedar policies to database filters, setting the foundation for future policy filter optimizations. The main goal is to solve lifetime issues that were blocking OR-to-IN clause optimizations in policy-based database queries.

## 🔍 What does this change?

- **Refactor PolicyComponents structure**: Separate raw policy storage from PolicySet creation
  - `PolicyComponents` now stores `Vec<Policy>` directly instead of pre-built `PolicySet`
  - Add lazy `create_policy_set()` method that builds PolicySet only when needed for Cedar evaluation
  - Update all `PolicySet` usage sites to use the new lazy creation pattern

- **Add direct policy-to-filter conversion**: New `Filter::for_policies()` method
  - Converts policy constraints directly to database filter expressions
  - Handles permit/forbid logic correctly with proper boolean algebra
  - Supports complex policy combinations (multiple permits, forbids, blank policies)

- **Comprehensive test coverage**: Add extensive test suite for policy conversion scenarios
  - Test single permits and forbids
  - Test multiple permits (the target case for future OR-to-IN optimization)
  - Test mixed permit/forbid combinations
  - Test blank permit/forbid edge cases

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

None. This is a refactoring that maintains existing behavior while enabling future optimizations.

## 🐾 Next steps

- Implement optimization analyzer to detect `OR`-to-`IN` optimization opportunities
- Add optimization data storage to `PolicyComponents`
- Implement optimized filter generation using `IN` clauses for multiple entity permits

## 🛡 What tests cover this?

- All existing policy evaluation tests continue to pass
- New comprehensive test suite in `libs/@local/graph/store/src/filter/mod.rs` covering:
  - Single policy conversions (permit/forbid)
  - Multiple policy scenarios
  - Complex permit/forbid combinations
  - Edge cases with blank policies

## ❓ How to test this?

1. Checkout the branch
2. Run `cargo test --package hash-graph-store --lib filter::tests::policy_conversion`
3. Confirm all policy conversion tests pass
4. Run existing integration tests to ensure no regressions

## 📹 Demo

This is an internal refactoring - no user-facing changes to demonstrate.

### CODE REVIEW

File: libs/@local/graph/authorization/src/policies/components.rs [hunk 1]
```diff
@@ -25,6 +26,22 @@ impl PolicyComponents {
...
+
+    /// Creates a [`PolicySet`] for Cedar evaluation.
+    ///
+    /// This is only needed when actually evaluating policies with Cedar.
+    /// For filter optimization, use the raw `policies` field directly.
+    ///
+    /// # Errors
+    ///
+    /// Returns error if policy set creation fails.
+    pub fn create_policy_set(
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/graph/authorization/src/policies/components.rs]: If you call `create_policy_set()` one would assume, that you _**create**_ a new policy set, not use a cached one.

File: libs/@local/graph/postgres-store/src/store/postgres/mod.rs [hunk 2]
```diff
@@ -268,7 +268,8 @@ where
         let web_id = WebId::new(parameter.id.unwrap_or_else(Uuid::new_v4));
 
         match policy_components
-            .policy_set
+            .create_policy_set()
```
[the remark this answers is not in the document]
Comment (hashdotai) [reply libs/@local/graph/postgres-store/src/store/postgres/mod.rs]: @copilot-pull-request-reviewer That's a good point. The repeated pattern of calling `create_policy_set()` followed by `change_context()` with different error types could be abstracted into a helper method or extension trait. Something like:

```rust
trait PolicyComponentsExt {
    fn create_policy_set_with_context<E>(&self, error_context: E) -> Result<PolicySet, Report<E>>;
}

impl PolicyComponentsExt for PolicyComponents {
    fn create_policy_set_with_context<E>(&self, error_context: E) -> Result<PolicySet, Report<E>> {
        self.create_policy_set().change_context(error_context)
    }
}
```

This would reduce the repetitive code and make it more maintainable. It could be especially useful if the error handling logic grows more complex in the future.

[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/graph/postgres-store/src/store/postgres/mod.rs]: What precisely you would move out?

File: libs/@local/graph/store/src/filter/mod.rs [hunk 3]
```diff
@@ -325,6 +333,212 @@ impl<'p> Filter<'p, Entity> {
...
+                // We have permits and forbids, so we allow access if there is at least one permit
+                (true, true) => Self::All(vec![
+                    Self::Any(permits),
+                    Self::Not(Box::new(Self::Any(forbids))),
+                ]),
+            }
+        }
+    }
+
+    // #[must_use]
```
Comment (hashdotai) [inline libs/@local/graph/store/src/filter/mod.rs:479]: You have a commented-out `for_resource_constraints` implementation that spans 61 lines (174-236). Since this is commented code and doesn't appear to be used, it should be removed entirely rather than being committed. Commented-out code can create confusion about what functionality is active and adds noise to the codebase.

File: libs/@local/graph/store/src/filter/mod.rs [hunk 4]
```diff
@@ -729,4 +943,255 @@
...
+                ),
+                create_test_policy(
+                    Effect::Permit,
+                    Some(ResourceConstraint::Entity(
+                        EntityResourceConstraint::Exact { id: entity_uuid_3 },
+                    )),
+                ),
+            ];
+
+            let filter =
```
Comment (hashdotai) [inline libs/@local/graph/store/src/filter/mod.rs:1074]: Consider adding explicit type annotation here to improve readability, especially since this is a complex return type:

```suggestion
let filter: Filter<Entity> = 
    Filter::<Entity>::for_policies(policies.iter().map(policy_to_tuple), actor_id);
```

File: libs/@local/graph/authorization/src/policies/components.rs [hunk 5]
```diff
@@ -16,7 +16,8 @@ use super::{
 #[derive(Debug)]
 pub struct PolicyComponents {
     pub actor_id: Option<ActorId>,
-    pub policy_set: PolicySet,
+    pub policies: Vec<Policy>,
+    pub tracked_actions: HashSet<ActionName>,
```
Comment (hashdotai) [inline libs/@local/graph/authorization/src/policies/components.rs:20]: The tracked_actions field is added in the `PolicyComponents` struct but it's not documented with a doc comment. Consider adding a doc comment to explain its purpose and usage pattern.

File: libs/@local/graph/postgres-store/src/store/postgres/knowledge/entity/mod.rs [hunk 6]
```diff
@@ -992,7 +992,8 @@ where
         let mut forbidden_instantiations = Vec::new();
         for entity_type_id in &entity_type_id_set {
             match policy_components
-                .policy_set
+                .create_policy_set()
+                .change_context(InsertionError)?
```
Comment (hashdotai) [inline libs/@local/graph/postgres-store/src/store/postgres/knowledge/entity/mod.rs:996]: For better error handling, consider adding a specific error message when changing the context here. This would help with debugging if policy set creation fails:

```suggestion
            .change_context(InsertionError)?
```

File: libs/@local/graph/postgres-store/src/store/postgres/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/graph/postgres-store/src/store/postgres/mod.rs]: Sorry, but this is stupid.

**Review by indietyp — APPROVED** [review]
